Handle string literal annotations

taion opened the pull request:
Fixes #290

taion commented on pyflakes/checker.py:
```diff
@@ -1160,9 +1166,11 @@ def addArgs(arglist):
                     self.report(messages.DuplicateArgument, node, arg)
 
-        for child in annotations + defaults:
-            if child:
```
I think this was redundant. The check was added in https://github.com/PyCQA/pyflakes/commit/34fd8ba800d288a41bde4c77d6be61effe94e177#diff-f0429ba5bb9de6e83abcf687aca4c012R751, but `handleNode` already checks for `node is None`, and I don't see how `child` can be a falsy value other than `None` here.

The previous revision did not have this extra falsiness check on `defaults`: https://github.com/PyCQA/pyflakes/commit/34fd8ba800d288a41bde4c77d6be61effe94e177#diff-f0429ba5bb9de6e83abcf687aca4c012L750.

taion commented:
Hold on, this breaks on http://mypy.readthedocs.io/en/latest/kinds_of_types.html#class-name-forward-references.

taion commented:
This is ready to go now.

taion commented:
Sorry for the repeated updates. This is now fixed.

myint commented:
Thanks!

I was testing this against the CPython standard library and it seemed to crash when run against one of the test files ([`test_functools.py`](https://github.com/python/cpython/blob/bdb8315c21825487b54852ff0511fb4881ea2181/Lib/test/test_functools.py)). You can reproduce it by doing the above or more specifically:

```
$ wget https://raw.githubusercontent.com/python/cpython/bdb8315c21825487b54852ff0511fb4881ea2181/Lib/test/test_functools.py

$ ./py/bin/pyflakes test_functools.py
Traceback (most recent call last):
  File "./py/bin/pyflakes", line 11, in <module>
    load_entry_point('pyflakes==1.6.0', 'console_scripts', 'pyflakes')()
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/api.py", line 208, in main
    warnings = checkRecursive(args, reporter)
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/api.py", line 165, in checkRecursive
    warnings += checkPath(sourcePath, reporter)
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/api.py", line 112, in checkPath
    return check(codestr, filename, reporter)
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/api.py", line 73, in check
    w = checker.Checker(tree, filename)
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/checker.py", line 498, in __init__
    self.runDeferred(self._deferredFunctions)
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/checker.py", line 535, in runDeferred
    handler()
  File "/Users/myint/tmp/pyflakes/py/lib/python3.6/site-packages/pyflakes/checker.py", line 929, in handleForwardAnnotation
    parsed = ast.parse(annotation.s).body[0].value
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 1
    This is a new annotation
                ^
SyntaxError: invalid syntax
```

taion commented:
Ah, my bad. Didn't handle annotations that weren't actually annotations. I've addressed it now and added test cases.

I've also added a new reported error here since there weren't any existing errors that covered this, and I don't think there are any reasons to want illegal annotations in normal code.

If you think this new error is reasonable, could you please advise on whether the naming is appropriate, given that this will be user-facing? The alternative would be to refer to this as a "string literal annotation" rather than a "forward annotation".

taion commented:
Rebased and squashed.

Can you confirm that you're okay with the naming and the semantics around the new error? Something like `a: 'A B'` or `a: 'A; B'` is legal Python, but I think it's vanishing unlikely to actually be intentionally used.

JackLeo commented:
when will this be released to pypi?

joar commented:
It would be super-useful to have this bugfix released, it's been several months since this was merged into master.

jcrada commented:
Hi, this bug was fixed in 2.0.0, but there is a case in which I still get `F401`. See below.

```python
import typing
if typing.TYPE_CHECKING:
    from .defuzzifier import Defuzzifier

self.defuzzifier: typing.Optional['Defuzzifier'] = None
```
`fuzzylite/variable.py:29:5: F401 '.defuzzifier.Defuzzifier' imported but unused`

jcrada commented:
I am not sure, but `mypy` complains if I write something that is not a forward annotation. For example, 
```python
self.defuzzifier: typing.Optional['Some comment'] = None
# mypy: fuzzylite/variable.py:126: error: syntax error in type comment

self.defuzzifier: typing.Optional['SomeComment'] = None
# fuzzylite/variable.py:126: error: Name 'SomeComment' is not defined
```

ambv commented:
@taion, in the type checker context, yes.

mgedmin commented:
I'm thinking if there should be an option to disable this, similarly how you can disable checking of code inside doctests.

(I'm currently working on a codebase that doesn't use mypy, but uses string annotations as documentation, without the corresponding `from typing import List` etc. imports, as those caused older pyflakes to emit warnings.  Now new pyflakes emits warnings for the annotations...)

taion commented:
@jcrada 

I'd suggest opening a new issue for that behavior if it's still causing you problems, though the default postponed evaluation of annotations from PEP 563 likely means that this shouldn't be a huge issue going forward.

The fix seems like it could be fairly straightforward – we track whether we're in an annotation context (perhaps with an `annotationDepth` attribute analogous to the current `nodeDepth` attribute), then handle all strings as postponed annotations rather than as just strings.

taion commented:
I was going to look at this at some point, but I haven't been able to find the time, and this isn't actively causing problems for me.

The lack of that behavior nevertheless does seem like a real bug, though.
